### ISSUE TITLE
default and automatic generated fields should be optional

### ISSUE THREAD

**Issue by MartinMuzatko** [CarterGrimmeisen/zod-prisma#96]
Hello there.
Whenever I use zod to validate my schema, i validate the data coming from userland. That means any HTTP/REST API, MQTT API etc. needs to get validated in order to create or update a database entry.

That means, all fields that are autogenerated (denoted by prisma with `@id`) should be marked as `optional` or removed from the model, as they are not needed neither in create or update mechanisms.

Fields that have a default value (denoted by `@default`) should be optional. This could be solved by making the field itself optional with `?`. However, I would like to disallow explicitly setting `null`, but allow implicitly accepting the default value by omitting the field.
This could be solved with `/// @zod.optional()` in the schema, but could also be a sensible default.

What are your thoughts? Is it worth to introduce an option for the generator? Am I maybe understanding something wrong?
In the end it is not so urgent, because it can be fixed on my end applying the correct types.
Thanks!

**Comment by CarterGrimmeisen** [comment]
No, you are understanding it correctly. I definitely agree this needs to be incorporated in some way. I'm currently brainstorming a more all-encompassing rework that should handle this, #93, and #90. So I'll update this issue when that becomes available! Thanks!

**Comment by davguij** [comment]
Hey @CarterGrimmeisen has there been any progress on this? Is this still on your radar? It would be an improvement I could surely use :)

**Comment by abenhamdine** [comment]
It doesn't make sense to me that a field could not be optional and have a default value.

> Fields that have a default value (denoted by @default) should be optional. This could be solved by making the field itself optional with ?. However, I would like to disallow explicitly setting null, but allow implicitly accepting the default value by omitting the field. 

Why would you want to disallow `null` and allow omitting the property, given than the final result will be the same (applying the default value in db) ? I fail to understand.

**Comment by MartinMuzatko** [comment]
I think you can forget about what I said about `null`. If a field has a default value (but is not optional by nature) it should be, as convention, use the default value if omitted.

I can think of at least two scenarios where I want to use a validation schema in a basic CRUD application.

1. Create - all required fields have to be required
2. Update - no fields are required

**Case 1 - Create**
How should fields behave that are required but have a default value? Are they really required? Should I be able to configure that?
Should the validator swallow a null?
**Example:**
``` prisma
model Notification {
  id        Int      @id @default(autoincrement())
  subject   String
  trigger   String?
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

`createdAt` and `updatedAt` are required fields as defined in the prisma schema, but they have a default value (`updatedAt` has implicitly a default value, but not denoted by `@default`)
`prisma.notification.create({ subject: 'whatever', })` will also not complain, if you omit these fields.
However, the zod validator does. 

If you create a `Notification`, you wouldn't want to validate for `createdAt` or `updatedAt` at best it should be disallowed to be set.

**Case 2 - Update**
`createdAt` and `updatedAt` is swollowed by prisma here anyway. Not allowed to overwrite.
On update, every field should be optional. So validation errors can only happen for provided fields that are of wrong type + other limitations like min/max etc.

**Comment by testerez** [comment]
Same issue here. I would be fine with a `/// @zod.ignore` special comment to remove any specific field from the generated schema. Also, a global ignore configuration would be nice to ignore fields for the whole schema. Something like:

```
generator zod {
  provider      = "zod-prisma"
  ignore = ["createdAt", "updatedAt"]
}
```

Would that make sense?

**Comment by testerez** [comment]
In case anyone is interested I added a feature on my fork to omit any field: https://github.com/testerez/zod-prisma#omit

You can install my fork with: `yarn add zod-prisma@testerez/zod-prisma`

I'd be happy to open a PR here but it looks like this project is not maintained anymore.
